File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/astar/Distance.java
package hu.bme.mit.theta.analysis.algorithm.cegar.astar;

public final class Distance implements Comparable<Distance> {
	private final Type type;
	private final int value;

	public enum Type {
		INFINITE,
		EXACT,
		UNKNOWN,
	}

	public Distance(Type type) {
		assert type != Type.EXACT;
		this.type = type;
		this.value = 0;
	}

	public Distance(Type type, int value) {
		this.type = type;
		this.value = value;
	}

	public Type getType() {
		return type;
	}

	public int getValue() {
		assert type == Type.EXACT;
		return value;
	}

	public boolean isKnown() {
		return type == Type.EXACT || type == Type.INFINITE;
	}

	// Both distances should be known
	@Override
	public int compareTo(Distance distance) {
		assert isKnown();
		assert distance.isKnown();
		if (getType() == Type.INFINITE && distance.getType() == Type.INFINITE) {
			return 0;
		}
		if (getType() == Type.INFINITE) {
			return 1;
		}
		if (distance.getType() == Type.INFINITE) {
			return -1;
		}
		int value1 = getValue();
		int value2 = distance.getValue();
		return value1 - value2;
	}

	@Override
	public boolean equals(Object object) {
		if (object.getClass() != Distance.class) return false;
		return compareTo((Distance) object) == 0;
	}

	@Override
	public String toString() {
		String result = "(";
		if (type == Type.EXACT) {
			result += "E," + value;
		} else if (type == Type.UNKNOWN) {
			result += "U";
		} else if (type == Type.INFINITE) {
			result += "I";
		}
		return result + ")";
	}
}

File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/astar/Distance.kt
package hu.bme.mit.theta.analysis.algorithm.cegar.astar

import hu.bme.mit.theta.analysis.algorithm.cegar.astar.strategy.HeuristicSearchType

class Distance private constructor(
	private val type: Type,
	value: Int,
) : Comparable<Distance> {
	val value = value
		get(): Int {
			// TODO if we auto set infinite for inf heuristic then propagateUpDistanceFromInfiniteDistance's 5) logic have to be rechecked
			check(hasValue)
			return field
		}

	val isFinite: Boolean
		get() = type == Type.FINITE

	val isInfinite: Boolean
		get() = type == Type.INFINITE

	val isKnown: Boolean
		get() = isFinite || isInfinite

	val isUnknown: Boolean
		get() = !isKnown

	val hasValue: Boolean
		get() = isFinite || isUnknown

	// Both distances should be known
	override fun compareTo(other: Distance): Int {
		return when {
			type == Type.INFINITE && other.type == Type.INFINITE -> 0
			type == Type.INFINITE -> 1
			other.type == Type.INFINITE -> -1
			else -> value.compareTo(other.value)
		}
	}

	override fun equals(other: Any?) = other is Distance && compareTo(other) == 0

	override fun hashCode() = 31 * type.hashCode() + value

	override fun toString() = when (type) {
		Type.FINITE -> "F$value"
		Type.INFINITE -> "I"
		Type.LOWERBOUND -> if (DI.heuristicSearchType == HeuristicSearchType.FULLY_ONDEMAND) "L$value" else "U"
	}

	enum class Type {
		FINITE,
		INFINITE,
		LOWERBOUND,
	}

	companion object {
		// There is no need to clear cache as distances to targets will keep get longer

		// static factory with caching: Distance object with same immutable content are often created
		private val finiteCache = HashMap<Int, Distance>(100) // random number, not measured
		fun finiteOf(value: Int) = finiteCache.computeIfAbsent(value) { Distance(Type.FINITE, it) }

		private val lowerBoundCache = HashMap<Int, Distance>(100)
		fun lowerBoundOf(value: Int) = lowerBoundCache.computeIfAbsent(value) { Distance(Type.LOWERBOUND, it) }
		fun lowerBoundOf(distance: Distance): Distance {
			// Lower-bound of infinite should be infinite
			check(distance.hasValue)
			return lowerBoundOf(distance.value)
		}

		fun sameTypeOfNoCache(distance: Distance, value: Int) = Distance(distance.type, value)

		val F0 = finiteOf(0)
		val INFINITE = Distance(Type.INFINITE, 0)
		val UNKNOWN = lowerBoundOf(0)
	}
}
